room/compiler/src/main/kotlin/androidx/room/processor/Context.kt

```diff
@@ -37,7 +37,8 @@ class Context private constructor(
     private val typeConverters: CustomConverterProcessor.ProcessResult,
     private val inheritedAdapterStore: TypeAdapterStore?,
     val cache: Cache,
-    private val canRewriteQueriesToDropUnusedColumns: Boolean
+    private val canRewriteQueriesToDropUnusedColumns: Boolean,
+    private val addTypeConverterFactory: (XType) -> Unit
```

Comment:
github doesn't let me comment on the previous discussion so i'll create a new one:
https://github.com/androidx/androidx/pull/75#discussion_r482967139

I'm still worried about this.
We keep type converters in Context because they are scoped to where they are declared and they are stateless.
The current implementation relies on a side effect which is a bit scary especially with caching where we try to avoid re-processing same classes (e.g. https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:room/compiler/src/main/kotlin/androidx/room/processor/TableEntityProcessor.kt;l=51)

I tried to create a repro for that but it didn't fail (not sure why, didn't try harder) but i'm afraid it might fail in similar cases.

In other words, there might be a failure for the following case:
There are two database, `Db1`,`Db2` that share same Entity `E1` where `E1` has a field that declares a type converter.

We start processing `Db1`, visit `E1` and add the factory.
We start processing `Db2`, visit `E1` again but it is cached hence return cached Entity information, missing on the factory.

I don't have a solution yet but the original implementation might be better than having a callback.
That is, have a mechanism for a `subContext` to report to its parent a type adapter.

Now, even that is not perfect though w/ the forks where sometimes we fork a context just to collect errors and sometime surpress them (https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:room/compiler/src/main/kotlin/androidx/room/processor/Context.kt;l=117?q=Context.kt)

Also, it would be great not to report errors on unused type converters but that is a good to have, not that important.

Reply:
I'm not super familiar with the codebase yet, not sure how exactly caching works but I think I know why it doesn't fail. One of the first things that `DatabaseProcessor` does is forking a baseContext. `fork()` method calls `CustomConverterProcessor.findConverters` before creating subcontext and returning it. So it happens before Daos and Entities are processed. That's why it works - Room executes `findConverters` logic without checking the cache.

Reply:
so the case above works because available converters is a cache key.

Reply:
> so the case above works because available converters is a cache key.

I'm trying to understand this, can you elaborate? Does that mean that E1 will be visited again when processing Db2 because converters is a cache key?

Reply:
One idea for gathering factories from lower levels (daos/entities etc) would be to modify fork method like this:
```
    fun fork(
        element: XElement,
        forceSuppressedWarnings: Set<Warning> = emptySet(),
        addTypeConverterFactory: ((XType) -> Unit)? = null // make it nullable
    ): Context {
        val suppressedWarnings = SuppressWarningProcessor.getSuppressedWarnings(element)
        val processConvertersResult =
            CustomConverterProcessor.findConverters(this, element, addTypeConverterFactory ?: this.addTypeConverterFactory) // pass argument if not null, otherwise use class property
        val canReUseAdapterStore = processConvertersResult.classes.isEmpty()
        // order here is important since the sub context should give priority to new converters.
        val subTypeConverters = if (canReUseAdapterStore) {
            this.typeConverters
        } else {
            processConvertersResult + this.typeConverters
        }
        val subSuppressedWarnings =
            forceSuppressedWarnings + suppressedWarnings + logger.suppressedWarnings
        val subCache = Cache(cache, subTypeConverters.classes, subSuppressedWarnings)
        val subCanRemoveUnusedColumns = canRewriteQueriesToDropUnusedColumns ||
                element.hasRemoveUnusedColumnsAnnotation()
        val subContext = Context(
                processingEnv = processingEnv,
                logger = RLog(logger.messager, subSuppressedWarnings, element),
                typeConverters = subTypeConverters,
                inheritedAdapterStore = if (canReUseAdapterStore) typeAdapterStore else null,
                cache = subCache,
                canRewriteQueriesToDropUnusedColumns = subCanRemoveUnusedColumns,
                addTypeConverterFactory = addTypeConverterFactory ?: this.addTypeConverterFactory) // same here, use arg if not null, otherwise pass class property
        subContext.databaseVerifier = databaseVerifier
        return subContext
    }
```
That way once a not null callback is passed to fork, it'll be used starting from that level and passed to lower levels. We're creating a proper callbacks that collects factories into a map inside `DatabaseProcessor` so all subcontexts would get this callback.

Reply:
OK here is a proposal.
tl;dr; Delegate verification to static methods in generated DAOs 🙈 

This provides the best behavior (we won't validate unless it is used) and does not require any change in the `Context`.

Right now, `CustomTypeConverterWrapper.typeConverterFactory` gets a field from the scope.
We can add a scope function specific to factories which can track all factories that are used.

`DaoWriter`, at the end, can generate a `static` method that returns the list of those factories.

`DatabaseWriter` can just query these static methods to build the list and verify against it. That even allows us to print a better runtime exception where we can tell the user "XXDAo needs this type converter factory".

wdyt?

Reply:
> > so the case above works because available converters is a cache key.
> 
> I'm trying to understand this, can you elaborate? Does that mean that E1 will be visited again when processing Db2 because converters is a cache key?

So E1 will be visited again for DB2 and then skipped because we've visited E1.
Right now it doesn't because our caching is not clever enough to understand that we've visited E1 before with same type converters hence it works fine.